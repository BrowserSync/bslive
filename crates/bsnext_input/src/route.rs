use std::collections::BTreeMap;
use std::fmt::{Display, Formatter};
use std::hash::{Hash, Hasher};
use std::ops::Deref;

use crate::watch_opts::WatchOpts;
use bsnext_resp::inject_opts::InjectOpts;

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub struct Route {
    pub path: String,
    #[serde(flatten)]
    pub kind: RouteKind,
    #[serde(flatten)]
    pub opts: Opts,
}

#[derive(Debug, Default, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub struct Opts {
    #[serde(flatten)]
    pub cors: Option<CorsOpts>,
    #[serde(flatten)]
    pub delay: Option<DelayOpts>,
    #[serde(default)]
    pub watch: WatchOpts,
    #[serde(default)]
    pub inject: InjectOpts,
    pub headers: Option<BTreeMap<String, String>>,
}

impl Default for Route {
    fn default() -> Self {
        Self {
            path: "/".to_string(),
            kind: RouteKind::new_html("default"),
            opts: Opts {
                headers: None,
                cors: None,
                delay: None,
                watch: Default::default(),
                inject: Default::default(),
            },
        }
    }
}

impl AsRef<Route> for Route {
    fn as_ref(&self) -> &Route {
        self
    }
}

impl Route {
    pub fn path(&self) -> &str {
        self.path.as_str()
    }
}

#[derive(Debug, Hash, PartialEq, Clone, serde::Deserialize, serde::Serialize)]
#[serde(untagged)]
pub enum RouteKind {
    Raw(RawRoute),
    Proxy(ProxyRoute),
    Dir(DirRoute),
}

impl RouteKind {
    pub fn new_html(html: impl Into<String>) -> Self {
        RouteKind::Raw(RawRoute::Html { html: html.into() })
    }
    pub fn new_raw(raw: impl Into<String>) -> Self {
        RouteKind::Raw(RawRoute::Raw { raw: raw.into() })
    }
    pub fn new_json(json: JsonWrapper) -> Self {
        RouteKind::Raw(RawRoute::Json { json })
    }
    pub fn new_sse(raw: impl Into<String>) -> Self {
        RouteKind::Raw(RawRoute::Sse { sse: raw.into() })
    }
}

#[derive(Debug, PartialEq, Clone, serde::Deserialize, serde::Serialize)]
pub struct JsonWrapper(pub serde_json::Value);

impl Display for JsonWrapper {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", serde_json::to_string(&self.0).expect("serde_json"))
    }
}

impl Deref for JsonWrapper {
    type Target = serde_json::Value;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl Hash for JsonWrapper {
    fn hash<H: Hasher>(&self, state: &mut H) {
        //todo: implement hashing for serde_value
        if let Ok(as_str) = serde_json::to_string(&self.0) {
            state.write(as_str.as_bytes());
        } else {
            todo!("handle error here?")
        }
    }
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub struct DirRoute {
    pub dir: String,
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub struct ProxyRoute {
    pub proxy: String,
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
#[serde(untagged)]
pub enum RawRoute {
    Html { html: String },
    Json { json: JsonWrapper },
    Raw { raw: String },
    Sse { sse: String },
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub enum CorsOpts {
    #[serde(rename = "cors")]
    Cors(bool),
}

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub enum DelayOpts {
    #[serde(rename = "delay")]
    Delay(DelayKind),
}

#[derive(
    Debug, Ord, PartialOrd, PartialEq, Eq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
pub enum DelayKind {
    #[serde(rename = "ms")]
    Ms(u64),
}

#[derive(
    Debug, Ord, PartialOrd, PartialEq, Eq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
pub enum DebounceDuration {
    #[serde(rename = "ms")]
    Ms(u64),
}

#[derive(
    Debug, Ord, PartialOrd, PartialEq, Eq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
#[serde(untagged)]
pub enum FilterKind {
    StringGlob(String),
    Extension { ext: String },
    Glob { glob: String },
    List(Vec<FilterKind>),
}

#[derive(
    Debug, Ord, PartialOrd, PartialEq, Eq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
pub struct Spec {
    #[serde(flatten)]
    pub opts: Option<SpecOpts>,
}

#[derive(
    Debug, Ord, PartialOrd, PartialEq, Eq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
pub struct SpecOpts {
    pub debounce: Option<DebounceDuration>,
    pub filter: Option<FilterKind>,
}

#[derive(
    Debug, PartialOrd, Ord, Eq, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize,
)]
pub struct Watcher {
    pub dir: String,
    pub debounce_ms: Option<u64>,
    pub filter: Option<FilterKind>,
}
