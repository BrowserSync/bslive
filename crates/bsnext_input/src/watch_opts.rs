use crate::route::Spec;

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
#[serde(untagged)]
pub enum WatchOpts {
    Bool(bool),
    InlineGlob(String),
    Spec(Spec),
}

impl Default for WatchOpts {
    fn default() -> Self {
        Self::Bool(true)
    }
}

impl WatchOpts {
    pub fn is_enabled(&self) -> bool {
        !matches!(self, WatchOpts::Bool(false))
    }
    pub fn spec(&self) -> Option<&Spec> {
        match self {
            WatchOpts::Bool(_) => None,
            WatchOpts::InlineGlob(_) => None,
            WatchOpts::Spec(spec) => Some(spec),
        }
    }
}
