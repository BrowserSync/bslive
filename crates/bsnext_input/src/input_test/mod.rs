use crate::route::{
    CorsOpts, DebounceDuration, DelayKind, DelayOpts, FilterKind, Route, Spec, SpecOpts, Watcher,
};
use crate::watch_opts::WatchOpts;
use crate::Input;

#[test]
fn test_deserialize() {
    let input = include_str!("../../../../examples/kitchen-sink/bslive.yml");
    let _: Input = serde_yaml::from_str(input).unwrap();
}

#[test]
fn test_deserialize_2() {
    #[derive(serde::Deserialize, serde::Serialize, Debug)]
    struct Config {
        pub items: Vec<Route>,
    }

    let input = r#"
    items:
      - path: /hello.js
        raw: "hello"
        cors: true
        delay:
            ms: 2000
      - path: /hello.js
        json: ["2", "3"]
      - path: /node_modules
        dir: ./node_modules
      - path: /node_modules
        dir: ./node_modules
      - path: /api
        proxy: example.com

        "#;
    let c: Config = serde_yaml::from_str(input).unwrap();
    let first = c.items.get(0).unwrap().to_owned();
    let opts = first.opts.cors.unwrap();
    assert_eq!(opts, CorsOpts::Cors(true));

    let delay_opts = first.opts.delay.unwrap();
    assert_eq!(delay_opts, DelayOpts::Delay(DelayKind::Ms(2000)));
}

#[test]
fn test_deserialize_cors_false() {
    #[derive(serde::Deserialize, serde::Serialize, Debug)]
    struct Config {
        pub items: Vec<Route>,
    }

    let input = r#"
    items:
      - path: /hello.js
        raw: "hello"
        cors: false
        "#;
    let c: Config = serde_yaml::from_str(input).unwrap();
    let first = c.items.get(0).unwrap().to_owned();
    let opts = first.opts.cors.unwrap();
    assert_eq!(opts, CorsOpts::Cors(false));
}

#[test]
fn test_deserialize_3_headers_control() {
    #[derive(serde::Deserialize, serde::Serialize, Debug)]
    struct Config {
        pub items: Vec<Route>,
    }

    let input = r#"
    items:
      - path: /api
        json: [1,2]
        "#;
    let c: Config = serde_yaml::from_str(input).unwrap();
    insta::assert_debug_snapshot!(c)
}

#[test]
fn test_deserialize_3_headers() {
    #[derive(serde::Deserialize, serde::Serialize, Debug)]
    struct Config {
        pub items: Vec<Route>,
    }

    let input = r#"
    items:
      - path: /api
        json: [1,2]
        headers:
            a: b
        "#;
    let c: Config = serde_yaml::from_str(input).unwrap();
    insta::assert_debug_snapshot!(c)
}

#[test]
fn test_deserialize_3() {
    let input = r#"
    path: /hello.js
    dir: "hello"
    cors: true
        "#;
    let _c: Route = serde_yaml::from_str(input).unwrap();
}

#[test]
fn test_deserialize_watch() {
    let input = r#"
    path: /hello.js
    dir: "hello"
    watch: true
        "#;
    let c: Route = serde_yaml::from_str(input).unwrap();
    assert_eq!(c.opts.watch, WatchOpts::Bool(true));
    let input = r#"
    path: /hello.js
    dir: "hello"
    watch: false
        "#;
    let c: Route = serde_yaml::from_str(input).unwrap();
    assert_eq!(c.opts.watch, WatchOpts::Bool(false));
    let input = r#"
    path: /hello.js
    dir: "hello"
    watch: "public/**/*.css"
        "#;
    let c: Route = serde_yaml::from_str(input).unwrap();
    assert_eq!(
        c.opts.watch,
        WatchOpts::InlineGlob("public/**/*.css".into())
    );
    let input = r#"
    path: /hello.js
    dir: "hello"
    watch:
      debounce:
        ms: 2000
        "#;
    let c: Route = serde_yaml::from_str(input).unwrap();
    assert_eq!(
        c.opts.watch,
        WatchOpts::Spec(Spec {
            opts: Some(SpecOpts {
                debounce: Some(DebounceDuration::Ms(2000)),
                filter: None,
            })
        })
    );
}

#[test]
fn test_deserialize_server_watch_list() {
    let input = r#"
servers:
- bind_address: 0.0.0.0:4000
  watchers:
    - dir: ./
    - dir: ./other
      debounce_ms: 2000
      filter:
        ext: "**/*.css"
"#;
    let c: Input = serde_yaml::from_str(input).unwrap();
    dbg!(&c);
    assert_eq!(
        c.servers.get(0).unwrap().watchers,
        vec![
            Watcher {
                dir: "./".to_string(),
                debounce_ms: None,
                filter: None,
            },
            Watcher {
                dir: "./other".to_string(),
                debounce_ms: Some(2000),
                filter: Some(FilterKind::Extension {
                    ext: "**/*.css".to_string()
                }),
            }
        ]
    )
}
