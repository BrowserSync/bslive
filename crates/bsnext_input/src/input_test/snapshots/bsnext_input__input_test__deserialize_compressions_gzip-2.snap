---
source: crates/bsnext_input/src/input_test/mod.rs
expression: c.get(1).unwrap().opts.compression
---
Some(
    CompType(
        Br,
    ),
)
